lib/auth/challenge_response.rb

```diff
@@ -1,12 +1,38 @@
 require_relative 'authentication_backend'
 
 class ChallengeResponse
-  attr_reader :response_type, :client_id, :session_id, :challenge_name, :challenge_parameters
+  attr_reader :email, :response_type, :client_id, :session_id,
+              :challenge_parameters, :secret_code, :flash_message
+  attr_accessor :challenge_name
 
-  def initialize(cognito_response:)
-    @response_type = AuthenticationBackend::CHALLENGE
+  def initialize(response_hash)
+    @response_type = response_hash[:type] || AuthenticationBackend::CHALLENGE
+    @email = response_hash[:email]
+    @secret_code = response_hash[:secret_code]
+    if response_hash[:cognito_response].nil?
+      retry_response(response_hash)
+    else
+      cognito_response(response_hash)
+    end
+  end
+
+  def to_h
+    instance_variables.map { |var| [var.to_s.delete('@').to_sym, instance_variable_get(var)] }.to_h
+  end
+
+private
+
+  def cognito_response(response_hash)
+    cognito_response = response_hash[:cognito_response]
     @challenge_name = cognito_response[:challenge_name]
     @session_id = cognito_response[:session]
     @challenge_parameters = cognito_response[:challenge_parameters]
   end
+
+  def retry_response(response_hash)
+    @flash_message = response_hash[:flash_message]
```

Comment:
why are we passing the flash_message as part of the response?

Reply:
I could call it message rather than flash message but the answer lies in the fact we can't access any of the flash message functionality in the devise authentication layer.  We want to capture why something failed so we can advise the user.  The simple things like their choice of password doesn't match criteria or the they entered didn't match the one Cognito was expecting.

Reply:
the fact is that we're in the model and not controller, so `flash` is not accessible. We should be using the built-in model errors... i.e. `errors.add`... but that requires further changes. We can leave this for now, but I'd like to come back to it and make sure we're consistent and doing it the right way.